File: src/lib/api/stock.ts
/**
 * Stock API 클라이언트 — 캔들, 기술적 지표, 현재가 조회.
 */

import { apiFetch } from "./client";

// ─── 타입 ───

export interface CandleData {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface CandleResponse {
  ticker: string;
  period: string;
  interval: string;
  candles: CandleData[];
  count: number;
}

export interface RSIData {
  value: number | null;
  signal: string;
}

export interface MACDData {
  macd_line: number | null;
  signal_line: number | null;
  histogram: number | null;
  signal: string;
}

export interface BollingerBands {
  upper: number | null;
  middle: number | null;
  lower: number | null;
  signal: string;
}

export interface SMAData {
  sma_20: number | null;
  sma_60: number | null;
  sma_120: number | null;
  signal: string;
}

export interface TechnicalIndicators {
  rsi: RSIData;
  macd: MACDData;
  bollinger_bands: BollingerBands;
  sma: SMAData;
}

export interface IndicatorResponse {
  ticker: string;
  indicators: TechnicalIndicators;
  calculated_at: string;
  data_points: number;
}

export interface StockQuote {
  ticker: string;
  price: number | null;
  change: number | null;
  change_percent: number | null;
  volume: number | null;
  name: string;
  fetched_at: string;
}

// ─── API 함수 ───

export async function fetchCandles(
  ticker: string,
  period = "6mo",
  interval = "1d",
): Promise<CandleResponse> {
  return apiFetch(
    `/stock/${encodeURIComponent(ticker)}/candles?period=${period}&interval=${interval}`,
  );
}

export async function fetchIndicators(
  ticker: string,
): Promise<IndicatorResponse> {
  return apiFetch(`/stock/${encodeURIComponent(ticker)}/indicators`);
}

export async function fetchQuote(ticker: string): Promise<StockQuote> {
  return apiFetch(`/stock/${encodeURIComponent(ticker)}/quote`);
}

interface QuoteResponse {
  quotes: StockQuote[];
  count: number;
}

export async function fetchMultipleQuotes(
  tickers: string[],
): Promise<StockQuote[]> {
  const res = await apiFetch<QuoteResponse>("/stock/quotes", {
    method: "POST",
    body: JSON.stringify({ tickers }),
  });
  return res.quotes;
}

File: backend/app/models/stock.py
"""주식 데이터 + 기술적 지표 Pydantic 모델."""

from datetime import datetime

from pydantic import BaseModel


# --- 캔들 데이터 ---


class CandleData(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int


class CandleResponse(BaseModel):
    ticker: str
    period: str
    interval: str
    candles: list[CandleData]
    count: int


# --- 기술적 지표 ---


class RSIData(BaseModel):
    value: float | None = None
    signal: str = ""


class MACDData(BaseModel):
    macd_line: float | None = None
    signal_line: float | None = None
    histogram: float | None = None
    signal: str = ""


class BollingerBands(BaseModel):
    upper: float | None = None
    middle: float | None = None
    lower: float | None = None
    signal: str = ""


class SMAData(BaseModel):
    sma_20: float | None = None
    sma_60: float | None = None
    sma_120: float | None = None
    signal: str = ""


class TechnicalIndicators(BaseModel):
    rsi: RSIData = RSIData()
    macd: MACDData = MACDData()
    bollinger_bands: BollingerBands = BollingerBands()
    sma: SMAData = SMAData()


class IndicatorResponse(BaseModel):
    ticker: str
    indicators: TechnicalIndicators
    calculated_at: datetime
    data_points: int


# --- 현재가 ---


class StockQuote(BaseModel):
    ticker: str
    price: float | None = None
    change: float | None = None
    change_percent: float | None = None
    volume: int | None = None
    name: str = ""
    fetched_at: datetime


class QuoteResponse(BaseModel):
    quotes: list[StockQuote]
    count: int
